File: sem1/Programming Fundamentals/Gomoku/repository/repository.py
from errors.errors import RepoError


class Repository(object):

    def __init__(self):
        self.__elems = []

    def __len__(self):
        return len(self.__elems)

    def add(self, elem):
        """
        Function that adds an element to the repository.
        :return: added element
                 -1 if error
        """

        if elem in self.__elems:
            raise RepoError("Element already exists!")

        self.__elems.append(elem)

        return elem

    def remove(self, elem):
        """
        Function that removes an element from the repository.
        :return: removed element
                 -1 if error
        """

        if elem not in self.__elems:
            raise RepoError("Inexisting element!")

        for i in range(len(self.__elems)):
            if self.__elems[i] == elem:
                elem = self.__elems[i]
                del self.__elems[i]
                return elem

        return -1

    def search(self, elem):
        """
        Function that searches for given element in the repository.
        :return: searched elem
                 -1 if element is not found
        """
        for e in self.__elems:
            if e == elem:
                return e

        return -1

    def getAll(self):
        """
        Function that returns all elements from the repository.
        :return: list of all elements
        """
        return self.__elems

    def removeAll(self):
        """
        Function that removes all elements from the repository.
        """
        self.__elems = []

File: sem2/Object Oriented Programming/LocalMovieDatabase/Repository.cpp
#include "Repository.h"



using namespace std;

// Default constructor for Repository
Repository::Repository(int capacity)
{
}

// Copy constructor for Repository
Repository::Repository(const Repository& v)
{	
	this->dvector = v.dvector;
}

// Initialize repo with existing vector
Repository::Repository(std::vector<Movie> v)
{
	this->dvector = v;
}

// Assignment operator for Repository
Repository& Repository::operator=(const Repository& v)
{
	if (this == &v)
		return *this;

	this->dvector = v.dvector;

	return *this;
}

/*
	Add an element to the Repository.
	Input: e - Movie
	Output: true, if the element was added
			false, if the element already exists
*/
bool Repository::add(const Movie& e)
{	
	auto it = std::find(this->dvector.begin(), this->dvector.end(), e);

	if (it != this->dvector.end())
		throw std::runtime_error("Element already exists in repository!\n");

	this->dvector.push_back(e);

	return true;
}


/*
	Remove an element from the Repository.
	Input: e - Movie
	Precond: e has the id of the element to be removed
	Output: true, if the element was removed
			false, if the element did not exist
*/
bool Repository::remove(const Movie e)
{
	auto it = std::find(this->dvector.begin(), this->dvector.end(), e);

	if (it != this->dvector.end()) {
		this->dvector.erase(it);
		return true;
	}
	
	throw std::runtime_error("Element was not found in repository!");
}

/*
	Update an element from the Repository.
	Input: e - Movie
	Precond: e has the id of the element to be updated
			 e contains the updated fields
	Output: true, if the element was updated
			false, if the element did not exist
*/
bool Repository::update(const Movie e)
{
	for(unsigned i=0; i<this->dvector.size(); i++)
		if (this->dvector[i] == e) {
			this->dvector[i] = e;
			return true;
		}

	throw std::runtime_error("Element was not found in repository!");
}

/*
	Search for an element in the Repository.
	Input: e - Movie
	Precond: e has the id of searched element
	Output: position of Movie, if the Movie was found in the Repository
			-1, otherwise
*/
int Repository::search(const Movie e)
{
	for (unsigned i = 0; i < this->dvector.size(); i++)
		if (this->dvector[i] == e)
			return i;

	return -1;
}

// Get size of the Repository
int Repository::getSize() const
{
	return this->dvector.size();
}


// Get element at given position
Movie Repository::getElem(int position) const {
	return this->dvector[position];
}



std::string Repository::getElemStr(int position) {
	std::string str = "";
	Movie e=this->dvector[position];
	str += "Title: " + e.getTitle() + "\tGenre: " + e.getGenre() + "\tYear: " + std::to_string(e.getYear()) + "\t Likes: " + std::to_string(e.getLikes()) + "\tURL: " + e.getTrailer();

	return str;
}

// Get all elements of the Repository represeted as string
std::string Repository::getAllString()
{
	std::string repoStr="";

	for (unsigned i = 0; i < this->dvector.size(); i++)
		repoStr += this->getElemStr(i) + "\n";

	return repoStr;
}

// Get all elements of the Repository
std::vector<Movie> Repository::getAll()
{
	std::vector<Movie> movies;

	movies = this->dvector;

	return movies;
}



Repository& operator+(Repository& repo, Movie elem) 
{
	repo.add(elem);
	return repo;
}

Repository& operator+(Movie elem, Repository& repo) 
{
	repo.add(elem);
	return repo;
}

